klippy/extras/safe_z_home.py

```diff
@@ -0,0 +1,85 @@
+import logging
+
+class SafeZHoming:
+    def __init__(self, config):
+        self.printer = config.get_printer()
+        try:
+            (self.home_x_pos,
+            self.home_y_pos) = config.get("home_xy_position",
+                                            default=",").split(',')
+        except:
+            raise config.error("Unable to parse home_xy_position in %s" % (
+                config.get_name()))
+
+        self.z_hop = config.getfloat("z_hop", default=0.0)
+        self.z_hop_speed = config.getfloat('z_hop_speed', 15., above=0.)
+        self.speed = config.getfloat('speed', 50.0, above=0.)
+        self.gcode = self.printer.lookup_object('gcode')
+        self.toolhead = None
+        self.kinematics = None
+        self.gcode.register_command("G28", None)
+        self.gcode.register_command("G28", self.cmd_G28)
+
+        if self.printer.lookup_object("homing_override", default=None):
+            raise config.error("homing_override and safe_z_homing cannot"
+                                    +" be used simultaneously")
+
+    def cmd_G28(self, params):
+        if not self.toolhead:
+            self.toolhead = self.printer.lookup_object('toolhead')
+            self.kinematics = self.toolhead.get_kinematics()
+
+        # Perform Z Hop if necessary
+        if self.z_hop != 0.0:
+            # Check if the zhop would exceed the printer limits
+            pos = self.toolhead.get_position()
+            kin_status = self.kinematics.get_status()
+            if (kin_status['homed'][-1] and
+                pos[2] + self.z_hop > kin_status['limits'][-1][1]):
+                self.gcode.respond_info(
+                    "No zhop performed, target Z out of bounds: " +
+                    str(pos[2] + self.z_hop)
+                )
+            else:
+                # Perform the Z-Hop
+                self.toolhead.set_position(pos, homing_axes=[2])
+                pos[2] = pos[2] + self.z_hop
+                self.toolhead.move(pos, self.z_hop_speed)
+                self.gcode.reset_last_position()
+
+        # Determine which axes we need to home
+        if not any([axis in params.keys() for axis in ['X', 'Y', 'Z']]):
+            need_x, need_y, need_z = [True] * 3
+        else:
+            need_x, need_y, need_z = tuple(axis in params
+                                           for axis in ['X', 'Y', 'Z'])
+
+        # Home XY axes if necessary
+        new_params = {}
+        if need_x:
+            new_params['X'] = '0'
+        if need_y:
+            new_params['Y'] = '0'
+        if new_params:
+            self.gcode.cmd_G28(new_params)
+        # Home Z axis if necessary
+        if need_z:
+            # Move to safe XY homing position
+            pos = self.toolhead.get_position()
+            if self.home_x_pos:
+                pos[0] = float(self.home_x_pos)
+            if self.home_y_pos:
+                pos[1] = float(self.home_y_pos)
+            self.toolhead.move(pos, self.speed)
```

Comment:
If a toolhead.move() call is made, self.gcode.reset_last_position() should really be invoked afterwards.